milli/src/update/index_documents/transform.rs

```diff
@@ -380,41 +390,111 @@ impl<'a, 'i> Transform<'a, 'i> {
-            doc.insert(key.to_string(), value);
+        // we recreate a json containing only the fields that needs to be flattened.
+        // all the raw values get inserted directly in the `key_value` vec.
+        for (key, value) in obkv.iter() {
+            if json_depth_checker::should_flatten_from_unchecked_slice(value) {
+                let key =
+                    mapping.get(&key).ok_or(InternalError::FieldIdMappingMissingEntry { key })?;
+                let key =
+                    self.fields_ids_map.name(*key).ok_or(FieldIdMapMissingEntry::FieldId {
+                        field_id: *key,
+                        process: "Flatten from field mapping.",
+                    })?;
+                let value = serde_json::from_slice::<serde_json::Value>(value)
+                    .map_err(InternalError::SerdeJson)?;
+                doc.insert(key.to_string(), value);
+            } else {
+                key_value.push((key, value.into()));
+            }
         }
 
+        println!("Flattening\n{}", serde_json::to_string_pretty(&doc).unwrap());
         let flattened = flatten_serde_json::flatten(&doc);
+        println!("Got\n{}", serde_json::to_string_pretty(&flattened).unwrap());
 
-        // Once we have the flattened version we can convert it back to obkv and
-        // insert all the new generated fields_ids (if any) in the fields ids map.
-        output_buffer.clear();
-        let mut writer = KvWriter::new(output_buffer);
-        let mut flattened: Vec<_> = flattened.into_iter().collect();
-        // we reorder the field to get all the known field first
-        flattened
-            .sort_unstable_by_key(|(key, _)| self.fields_ids_map.id(&key).unwrap_or(FieldId::MAX));
-
-        for (key, value) in flattened {
+        // Once we have the flattened version we insert all the new generated fields_ids
+        // (if any) in the fields ids map and serialize the value.
+        for (key, value) in flattened.into_iter() {
             let fid = self.fields_ids_map.insert(&key).ok_or(UserError::AttributeLimitReached)?;
             let value = serde_json::to_vec(&value).map_err(InternalError::SerdeJson)?;
-            writer.insert(fid, &value)?;
+            key_value.push((fid, value.clone().into()));
+
             if field_buffer_cache.iter().find(|(id, _)| *id == fid).is_none() {
                 field_buffer_cache.push((fid, value.into()));
             }
         }
 
+        // we sort the key. If there was a conflict between the obkv and the new generated value the
+        // keys will be consecutive.
+        key_value.sort_unstable_by(|(a, _), (b, _)| a.cmp(b));
+
+        Self::create_obkv_from_key_value(&mut key_value, output_buffer)?;
+        Ok(())
+    }
+
+    /// Generate an obkv from a slice of key / value sorted by key.
+    fn create_obkv_from_key_value(
+        key_value: &mut [(FieldId, Cow<[u8]>)],
+        output_buffer: &mut Vec<u8>,
+    ) -> Result<()> {
```

Comment:
Could you debug assert that the key_value array is ordered by `FieldId`.

Reply:
Oh that's super smart thanks!